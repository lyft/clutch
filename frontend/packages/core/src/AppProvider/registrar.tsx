import type { UserConfiguration } from ".";
import type { ConfiguredRoute, Workflow, WorkflowConfiguration } from "./workflow";

const workflowRoutes = (
  workflowId: string,
  workflow: WorkflowConfiguration,
  configuration: UserConfiguration
): ConfiguredRoute[] => {
  const workflowConfig = configuration?.[workflowId] || {};
  const allRoutes = Object.keys(workflowConfig).map(key => {
    return {
      ...workflow.routes[key],
      ...workflowConfig[key],
    };
  });

  const validRoutes = allRoutes.filter(route => {
    const requiredRouteProps = route?.requiredConfigProps || [];
    const missingProps = requiredRouteProps.filter((prop: string) => {
      return route.componentProps?.[prop] === undefined;
    });

    const isValidRoute = missingProps.length === 0;
    if (!isValidRoute) {
      /* eslint-disable-next-line no-console */
      console.warn(
        `[${workflowId}][${route.path}] Not registered: Invalid config - missing required component props ${missingProps}`
      );
    }
    return isValidRoute;
  });

  return validRoutes;
};

/**
 * Determine all user registered workflows on the application and apply filters, if any.
 * @param workflows a map of workflow keys to functions that return their configuration.
 * @param configuration the user configuration, usually read in from the clutch.config.js file.
 * @param filters a list of filters to apply to the user registered workflows.
 * @returns
 */
const registeredWorkflows = async (
  workflows: { [key: string]: () => WorkflowConfiguration },
  configuration: UserConfiguration,
  filters: { (workflows: Workflow[]): Promise<Workflow[]> }[] = []
): Promise<Workflow[]> => {
  let validWorkflows = Object.keys(workflows || [])
    .map((workflowId: string) => {
      const workflow = workflows[workflowId]();
      try {
        return { ...workflow, routes: workflowRoutes(workflowId, workflow, configuration) };
      } catch {
        // n.b. if the routes aren't configured properly we drop the workflow
        /* eslint-disable-next-line no-console */
        console.warn(
          `Skipping registration of ${workflowId || "unknown"} workflow due to invalid config`
        );
        return null;
      }
    })
    .filter(workflow => workflow !== null);
  filters.forEach(f => {
    f(validWorkflows).then(w => {
      validWorkflows = w;
    });
  });
  return validWorkflows;
};

export default registeredWorkflows;
