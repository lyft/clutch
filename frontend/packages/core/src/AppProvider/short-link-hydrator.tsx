import React from "react";
import type { clutch as IClutch } from "@clutch-sh/api";

import { useShortLinkContext } from "../Contexts/shortlink-context";
import { WorkflowStorageContext } from "../Contexts/workflow-storage-context";
import { retrieveData } from "../Contexts/workflow-storage-context/helpers";
import workflowStorageContextReducer from "../Contexts/workflow-storage-context/reducer";
import type { WorkflowStorageContextProps } from "../Contexts/workflow-storage-context/types";
import { defaultWorkflowStorageState } from "../Contexts/workflow-storage-context/types";
import { Toast } from "../Feedback";

interface ShortLinkHydratorProps {
  hydrate: () => IClutch.shortlink.v1.IShareableState[] | null;
  onClear: () => void;
  children: React.ReactElement;
}

/**
 * Hydrator which is a wrapper for workflows
 * Will check on load if there exists any hydrated data for the current workflow
 * If there is it will populate the state and provide a toast
 */
const ShortLinkHydrator = ({
  hydrate,
  onClear,
  children,
}: ShortLinkHydratorProps): React.ReactElement => {
  const { storeData } = useShortLinkContext();
  const [workflowStorageState, dispatch] = React.useReducer(
    workflowStorageContextReducer,
    defaultWorkflowStorageState
  );

  React.useEffect(() => {
    const data = hydrate();

    if (data) {
      dispatch({ type: "HYDRATE", payload: { data } });
      onClear();
    }
  }, []);

  React.useEffect(() => {
    if (workflowStorageState.tempStore) {
      storeData(workflowStorageState.tempStore);
    }
  }, [workflowStorageState]);

  const workflowStorageProviderProps: WorkflowStorageContextProps = {
    shortLinked: workflowStorageState.shortLinked,
    storeData: (componentName: string, key: string, data: any, localStorage?: boolean) =>
      dispatch({ type: "STORE_DATA", payload: { componentName, key, data, localStorage } }),
    removeData: (componentName: string, key: string, localStorage?: boolean) =>
      dispatch({ type: "REMOVE_DATA", payload: { componentName, key, localStorage } }),
    retrieveData: (componentName: string, key: string, defaultData?: any) =>
      retrieveData(workflowStorageState.store, componentName, key, defaultData),
  };

  return (
    <WorkflowStorageContext.Provider value={workflowStorageProviderProps}>
      {workflowStorageState.shortLinked && (
        <Toast title="Visited Short Link">Local Workflow Data will not be saved until reload</Toast>
      )}
      {children}
    </WorkflowStorageContext.Provider>
  );
};

export default ShortLinkHydrator;
