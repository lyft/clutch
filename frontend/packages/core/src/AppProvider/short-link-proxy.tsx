import React from "react";
import { useParams } from "react-router-dom";
import type { clutch as IClutch } from "@clutch-sh/api";

import { useNavigate } from "../navigation";
import { client } from "../Network";
import type { ClutchError } from "../Network/errors";

/**
 * The base for a short link route
 */
export const ShortLinkBaseRoute = "goto";

/**
 * Will return a ShortLink route
 * @param origin the windows origin
 * @param hash the hash to use for the short link
 * @returns string
 */
export const generateShortLinkRoute = (origin: string, hash: string) =>
  `${origin}/${ShortLinkBaseRoute}/${hash}`;

interface ShortLinkProps {
  hydrate: (data: IClutch.shortlink.v1.IShareableState[]) => void;
  onError: (error: ClutchError) => void;
  setLoading: (loading: boolean) => void;
}

const fetchData = async (hash, hydrate, onError) => {
  const requestData: IClutch.shortlink.v1.IGetRequest = { hash };

  return client
    .post("/v1/shortlink/get", requestData)
    .then(response => {
      const { path, state } = response.data as IClutch.shortlink.v1.IGetResponse;

      hydrate(state);
      return path;
    })
    .catch((error: ClutchError) => {
      onError(error);
      return "/";
    });
};

/**
 * Component that will be present for a route which will look for a short link hash
 * If found
 * - It will set a loading state
 * - Then it will call down to the API with the hash and ask for any data pertaining to it
 * - If the API call is successful
 *   - It will use the given hydrate function to send the returned state off to the StorageContext
 *   - It will navigate to the route given in the returned state
 * - If the API call is not successful
 *   - It will leave a warning message in the console
 *   - Then navigate back to the home page
 * - Then it will remove the loading state
 */
const ShortLinkProxy = ({ hydrate, onError, setLoading }: ShortLinkProps) => {
  const { hash } = useParams();
  const navigate = useNavigate();

  React.useEffect(() => {
    if (hash) {
      setLoading(true);
      (async function loadSL() {
        const path = await fetchData(hash, hydrate, onError);

        navigate(path);
        setLoading(false);
      })();
    }
  }, [hash]);

  // currently return null so that nothing is rendered
  return null;
};

export default ShortLinkProxy;
