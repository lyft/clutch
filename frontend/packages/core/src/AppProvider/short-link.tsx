import React from "react";
import { useParams } from "react-router-dom";
import type { clutch as IClutch } from "@clutch-sh/api";

import { useNavigate } from "../navigation";
import { client } from "../Network";
import type { ClutchError } from "../Network/errors";

interface ShortLinkProps {
  hydrate: (data: IClutch.shortlink.v1.IShareableState[], route: string) => void;
  onError: (error: string) => void;
}

/**
 * Component that will be present for a route which will look for a short link hash
 * - If found
 * - It will call down to the API with the hash and ask for any data pertaining to it
 * - If the API call is successful
 *   - It will use the given hydrate function to send the returned state off to the StorageContext
 *   - It will navigate to the route given in the returned state
 * - If the API call is not successful
 *   - It will leave a warning message in the console
 *   - Then navigate back to the home page
 */
const ShortLink = ({ hydrate, onError }: ShortLinkProps) => {
  const { hash } = useParams();
  const navigate = useNavigate();

  React.useEffect(() => {
    if (hash) {
      const requestData: IClutch.shortlink.v1.IGetRequest = { hash };

      client
        .post("/v1/shortlink/get", requestData)
        .then(response => {
          const { path = "/", state } = response.data as IClutch.shortlink.v1.IGetResponse;

          // we only want the pathname to match against so we pull off any search params for storage
          hydrate(state, path.split("?")[0]);
          navigate(path);
        })
        .catch((error: ClutchError) => {
          // eslint-disable-next-line no-console
          console.error(error?.status?.text);
          onError(`Unable to retrieve Shortlink '${hash}', redirecting home`);
          navigate("/");
        });
    }
  }, [hash]);

  // currently return null so that nothing is rendered
  // TODO: either make a loading spinner or something to go here if the API calls are not responsive enough
  return null;
};

export default ShortLink;
