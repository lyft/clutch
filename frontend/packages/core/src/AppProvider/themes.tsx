import React from "react";
import { ThemeProvider as EmotionThemeProvider } from "@emotion/react";
import { createMuiTheme, CssBaseline, MuiThemeProvider, ThemeOptions } from "@material-ui/core";
import { useTheme as useMuiTheme } from "@material-ui/core/styles";
import type { PaletteOptions } from "@material-ui/core/styles/createPalette";
import { StylesProvider } from "@material-ui/styles";
import { ThemeProvider as StyledThemeProvider } from "styled-components";

interface ClutchPalette extends PaletteOptions {
  accent: {
    main: string;
  };
  destructive: {
    main: string;
  };
}

interface ClutchTheme extends ThemeOptions {
  palette: ClutchPalette;
}

declare module "styled-components" {
  export interface ClutchTheme extends ThemeOptions {
    palette: ClutchPalette;
  }
}

const WHITE = "#ffffff";
const GRAY = "#D7DADB";
const TEAL = "#02acbe";
const RED = "#EF474D";
const NAVY = "#2D3F50";

const lightPalette = (): ClutchPalette => {
  return {
    accent: {
      main: TEAL,
    },
    destructive: {
      main: RED,
    },
    primary: {
      main: WHITE,
    },
    secondary: {
      main: NAVY,
    },
    background: {
      default: WHITE,
      paper: WHITE,
    },
    text: {
      primary: NAVY,
      secondary: GRAY,
    },
  };
};

const lightTheme = () => {
  return createMuiTheme({
    palette: lightPalette(),
    transitions: {
      // https://material-ui.com/getting-started/faq/#how-can-i-disable-transitions-globally
      create: () => "none"
    },
    props: {
      MuiButtonBase: {
        // https://material-ui.com/getting-started/faq/#how-can-i-disable-the-ripple-effect-globally
        disableRipple: true,
      },
    },
    overrides: {
      MuiTypography: {
        colorPrimary: {
          color: NAVY,
        },
        colorSecondary: {
          color: GRAY,
        },
      },
      MuiLink: {
        root: {
          color: TEAL,
        },
      },
    },
  });
};

const useTheme = () => {
  return useMuiTheme() as ClutchTheme;
};

interface ThemeProps {
  variant?: "light";
}

const Theme: React.FC<ThemeProps> = ({ children }) => {
  const theme = lightTheme;
  return (
    <MuiThemeProvider theme={theme()}>
      <EmotionThemeProvider theme={theme()}>
        <StyledThemeProvider theme={theme()}>
          <CssBaseline />
          <StylesProvider injectFirst>{children}</StylesProvider>
        </StyledThemeProvider>
      </EmotionThemeProvider>
    </MuiThemeProvider>
  );
};

export { Theme, useTheme };
