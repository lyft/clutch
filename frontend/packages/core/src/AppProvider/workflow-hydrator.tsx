import React from "react";
import type { clutch as IClutch } from "@clutch-sh/api";
import { Grid } from "@material-ui/core";

import { useShortLinkContext } from "../Contexts";
import { WorkflowStorageContext } from "../Contexts/workflow-storage-context";
import { retrieveData } from "../Contexts/workflow-storage-context/helpers";
import workflowStorageContextReducer from "../Contexts/workflow-storage-context/reducer";
import type { WorkflowStorageContextProps } from "../Contexts/workflow-storage-context/types";
import { defaultWorkflowStorageState } from "../Contexts/workflow-storage-context/types";
import { Alert } from "../Feedback";
import styled from "../styled";

interface WorkflowHydratorProps {
  children: React.ReactElement;
  hydrateData: () => IClutch.shortlink.v1.IShareableState[];
  onClear: () => void;
}

const StyledAlert = styled(Alert)({
  zIndex: 1200,
  position: "absolute",
  padding: "6px 8px",
});

/**
 * Hydrator which is a wrapper for workflows
 * Will check on load if there exists any hydrated data for the current workflow
 * If there is it will populate the state and provide an alert above the workflow
 */
const WorkflowHydrator = ({
  hydrateData,
  onClear,
  children,
}: WorkflowHydratorProps): React.ReactElement => {
  const { storeData } = useShortLinkContext();
  const [state, dispatch] = React.useReducer(
    workflowStorageContextReducer,
    defaultWorkflowStorageState
  );

  React.useEffect(() => {
    const hydratedData = hydrateData();
    if (hydratedData) {
      dispatch({ type: "HYDRATE", payload: { data: hydratedData } });
      onClear();
    }
  }, []);

  React.useEffect(() => {
    if (state.tempStore) {
      storeData(state.tempStore);
    }
  }, [state]);

  const storageProviderProps: WorkflowStorageContextProps = {
    shortLinked: state.shortLinked,
    storeData: (componentName: string, key: string, data: unknown, localStorage?: boolean) =>
      dispatch({ type: "STORE_DATA", payload: { componentName, key, data, localStorage } }),
    removeData: (componentName: string, key: string, localStorage?: boolean) =>
      dispatch({ type: "REMOVE_DATA", payload: { componentName, key, localStorage } }),
    retrieveData: (componentName: string, key: string, defaultData?: unknown) =>
      retrieveData(state.store, componentName, key, defaultData),
  };

  return (
    <WorkflowStorageContext.Provider value={storageProviderProps}>
      {state.shortLinked && (
        <Grid container direction="column" alignItems="flex-end">
          <StyledAlert severity="warning">
            Local Workflow Data will not be saved until reload
          </StyledAlert>
        </Grid>
      )}
      {children}
    </WorkflowStorageContext.Provider>
  );
};

export default WorkflowHydrator;
