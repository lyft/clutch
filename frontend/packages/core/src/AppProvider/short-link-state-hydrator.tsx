import React from "react";
import { Grid } from "@mui/material";

import { useShortLinkContext, WorkflowStorageContext } from "../Contexts";
import { retrieveData as retrieveDataHelper } from "../Contexts/workflow-storage-context/helpers";
import workflowStorageContextReducer from "../Contexts/workflow-storage-context/reducer";
import type {
  HydrateState,
  WorkflowStorageContextProps,
} from "../Contexts/workflow-storage-context/types";
import { defaultWorkflowStorageState } from "../Contexts/workflow-storage-context/types";
import { Alert } from "../Feedback";
import { Link } from "../link";
import styled from "../styled";

import { generateShortLinkRoute } from "./short-link-proxy";

interface ShortLinkStateHydratorProps {
  children: React.ReactElement;
  /** Data from ShortLink API to be hydrated into the  */
  sharedState: HydrateState;
  /** Used to clear temporary storage variable in the AppProvider */
  clearTemporaryState: () => void;
}

/** Allows the alert to float on top of all other components on the page */
const StyledAlert = styled(Alert)({
  zIndex: 1200,
  position: "absolute",
  padding: "6px 8px",
});

/**
 * Hydrator which is a wrapper for workflows
 * Will check on load if there exists any hydrated data for the current workflow
 * If there is it will populate the state and provide an alert above the workflow
 */
const ShortLinkStateHydrator = ({
  children,
  clearTemporaryState,
  sharedState,
}: ShortLinkStateHydratorProps): React.ReactElement => {
  const { storeWorkflowSession } = useShortLinkContext();
  const [state, dispatch] = React.useReducer(
    workflowStorageContextReducer,
    defaultWorkflowStorageState
  );

  React.useEffect(() => {
    if (sharedState?.state?.length) {
      dispatch({ type: "HYDRATE", payload: { data: sharedState } });
      clearTemporaryState();
    }
  }, [sharedState]);

  React.useEffect(() => {
    if (state.workflowSessionStore) {
      storeWorkflowSession(state.workflowSessionStore);
    }
  }, [state]);

  function retrieveData<T>(componentName: string, key: string, defaultData: T): T {
    return retrieveDataHelper(state.workflowStore, componentName, key, defaultData);
  }

  const storageProviderProps: WorkflowStorageContextProps = React.useMemo(
    () => ({
      /**
       * Boolean representing whether the state has been hydrated
       */
      fromShortLink: state.fromShortLink,
      /**
       * StoreData context function which will allow a component to write data for use in shortlinks as well as
       * store locally
       * @param componentName Name of the component that data is being stored under
       * @param key A lookup key used to reference the specific data set being stored
       * @param data The data being stored
       * @param localStorage Optional boolean on whether to write data to the localStorage as well
       * @returns void
       */
      storeData: (componentName: string, key: string, data: unknown, localStorage?: boolean) =>
        dispatch({ type: "STORE_DATA", payload: { componentName, key, data, localStorage } }),
      /**
       * RemoveData context function which will allow a component to remove data from use in shortlinks as well
       * locally if preferred
       * @param componentName Name of the component that data is being removed under
       * @param key A lookup key used to reference the specific data set being removed
       * @param localStorage Optional boolean on whether to remove data from localStorage as well
       * @returns
       */
      removeData: (componentName: string, key: string, localStorage?: boolean) =>
        dispatch({ type: "REMOVE_DATA", payload: { componentName, key, localStorage } }),
      /**
       * RetrieveData context function which will allow a component to retrieve data from a hydrated short link and
       * barring that will return data from local storage if available
       * @param componentName Name of the component that data is being retrieved under
       * @param key A lookup key used to reference the specific data set being retrieved
       * @param defaultData Optional set of data returned if nothing is found in the hydrator or localStorage
       * @returns
       */
      retrieveData,
    }),
    [state]
  );

  return (
    <WorkflowStorageContext.Provider value={storageProviderProps}>
      {state.fromShortLink && (
        <Grid container direction="column" alignItems="flex-end">
          <StyledAlert severity="warning">
            <div style={{ display: "flex" }}>
              Loaded shared state
              {state.hash && state.hash.length > 0 && (
                <>
                  &quot;
                  <Link href={generateShortLinkRoute(window.location.origin, state.hash)}>
                    {state.hash}
                  </Link>
                  &quot;
                </>
              )}
              . Any local changes will not be preserved.
            </div>
          </StyledAlert>
        </Grid>
      )}
      {children}
    </WorkflowStorageContext.Provider>
  );
};

export default ShortLinkStateHydrator;
