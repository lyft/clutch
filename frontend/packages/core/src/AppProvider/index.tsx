import React from "react";
import { BrowserRouter as Router, Outlet, Route, Routes } from "react-router-dom";
import _ from "lodash";

import AppLayout from "../AppLayout";
import { ApplicationContext } from "../Contexts/app-context";
import { FEATURE_FLAG_POLL_RATE, featureFlags } from "../flags";
import Landing from "../landing";
import NotFound from "../not-found";

import registeredWorkflows from "./registrar";
import { Theme } from "./themes";
import type { ConfiguredRoute, Workflow, WorkflowConfiguration } from "./workflow";
import ErrorBoundary from "./workflow";

export interface UserConfiguration {
  [packageName: string]: {
    [key: string]: ConfiguredRoute;
  };
}

/**
 * Filter workflow routes using available feature flags.
 * @param workflows a list of valid Workflow objects.
 */
const featureFlagFilter = (workflows: Workflow[]): Promise<Workflow[]> => {
  return featureFlags().then(flags =>
    workflows.filter(workflow => {
      /* eslint-disable-next-line no-param-reassign */
      workflow.routes = workflow.routes.filter(route => {
        const show =
          route.featureFlag === undefined ||
          (flags?.[route.featureFlag] !== undefined &&
            flags[route.featureFlag].booleanValue === true);
        return show;
      });
      return workflow.routes.length !== 0;
    })
  );
};

interface ClutchAppProps {
  availableWorkflows: {
    [key: string]: () => WorkflowConfiguration;
  };
  configuration: UserConfiguration;
}

const ClutchApp: React.FC<ClutchAppProps> = ({
  availableWorkflows,
  configuration: userConfiguration,
}) => {
  const [workflows, setWorkflows] = React.useState<Workflow[]>([]);
  const [isLoading, setIsLoading] = React.useState<boolean>(true);

  const loadWorkflows = () => {
    registeredWorkflows(availableWorkflows, userConfiguration, [featureFlagFilter]).then(w => {
      setWorkflows(w);
      setIsLoading(false);
    });
  };

  React.useEffect(() => {
    loadWorkflows();
    const interval = setInterval(loadWorkflows, FEATURE_FLAG_POLL_RATE);
    return () => clearInterval(interval);
  }, []);

  const [discoverableWorkflows, setDiscoverableWorkflows] = React.useState([]);
  React.useEffect(() => {
    /** Filter out all of the workflows that are configured to be `hideNav: true`.
     * This prevents the workflows from being discoverable by the user from the UI,
     * both search and drawer navigation.
     *
     * The routes for all configured workflows will still be reachable
     * by manually providing the full path in the URI.
     */
    const pw = _.cloneDeep(workflows).filter(workflow => {
      const publicRoutes = workflow.routes.filter(route => {
        return !(route?.hideNav !== undefined ? route.hideNav : false);
      });
      workflow.routes = publicRoutes; /* eslint-disable-line no-param-reassign */
      return publicRoutes.length !== 0;
    });
    setDiscoverableWorkflows(pw);
  }, [workflows]);

  return (
    <Router>
      <Theme variant="light">
        <div id="App">
          <ApplicationContext.Provider value={{ workflows: discoverableWorkflows }}>
            <Routes>
              <Route path="/*" element={<AppLayout isLoading={isLoading} />}>
                <Route key="landing" path="/" element={<Landing />} />
                {workflows.map((workflow: Workflow) => (
                  <ErrorBoundary workflow={workflow} key={workflow.path.split("/")[0]}>
                    <Route path={`${workflow.path}/*`} element={<Outlet />}>
                      {workflow.routes.map(route => {
                        const heading = route.displayName
                          ? `${workflow.displayName}: ${route.displayName}`
                          : workflow.displayName;
                        return (
                          <Route
                            key={workflow.path}
                            path={`${route.path}`}
                            element={React.cloneElement(<route.component />, {
                              ...route.componentProps,
                              heading,
                            })}
                          />
                        );
                      })}
                      <Route key={`${workflow.path}/notFound`} path="*" element={<NotFound />} />
                    </Route>
                  </ErrorBoundary>
                ))}
                <Route key="notFound" path="*" element={<NotFound />} />
              </Route>
            </Routes>
          </ApplicationContext.Provider>
        </div>
      </Theme>
    </Router>
  );
};

export default ClutchApp;
