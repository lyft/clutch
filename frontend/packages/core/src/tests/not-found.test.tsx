import React from "react";
import { mount } from "enzyme";
import { ThemeProvider as StyledThemeProvider } from "styled-components";

import { getTheme } from "../AppProvider/themes";
import NotFound from "../not-found";

describe("Not Found component", () => {
  let component;

  beforeAll(() => {
    component = mount(
      <StyledThemeProvider theme={getTheme()}>
        <NotFound />
      </StyledThemeProvider>
    );
  });

  it("renders correctly", () => {
    expect(component.debug()).toMatchSnapshot();
  });
});
