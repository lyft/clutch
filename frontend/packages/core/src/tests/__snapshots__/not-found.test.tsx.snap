// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Not Found component renders correctly 1`] = `
"<ThemeProvider theme={{...}}>
  <NotFound>
    <WithStyles(ForwardRef(Grid)) container={true} direction=\\"column\\" justify=\\"center\\" alignItems=\\"center\\">
      <ForwardRef(Grid) classes={{...}} container={true} direction=\\"column\\" justify=\\"center\\" alignItems=\\"center\\">
        <div className=\\"MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center\\">
          <Styled(WithStyles(ForwardRef(Grid))) item={true}>
            <WithStyles(ForwardRef(Grid)) item={true} className=\\"sc-AxjAm fOtbrD\\">
              <ForwardRef(Grid) classes={{...}} item={true} className=\\"sc-AxjAm fOtbrD\\">
                <div className=\\"MuiGrid-root sc-AxjAm fOtbrD MuiGrid-item\\">
                  <ForwardRef fontSize=\\"inherit\\">
                    <WithStyles(ForwardRef(SvgIcon)) fontSize=\\"inherit\\">
                      <ForwardRef(SvgIcon) classes={{...}} fontSize=\\"inherit\\">
                        <svg className=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeInherit\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" color={[undefined]} aria-hidden=\\"true\\" role=\\"presentation\\">
                          <path d=\\"M15 3H6c-.83 0-1.54.5-1.84 1.22l-3.02 7.05c-.09.23-.14.47-.14.73v2c0 1.1.9 2 2 2h6.31l-.95 4.57-.03.32c0 .41.17.79.44 1.06L9.83 23l6.59-6.59c.36-.36.58-.86.58-1.41V5c0-1.1-.9-2-2-2zm4 0v12h4V3h-4z\\" />
                        </svg>
                      </ForwardRef(SvgIcon)>
                    </WithStyles(ForwardRef(SvgIcon))>
                  </ForwardRef>
                </div>
              </ForwardRef(Grid)>
            </WithStyles(ForwardRef(Grid))>
          </Styled(WithStyles(ForwardRef(Grid)))>
          <WithStyles(ForwardRef(Grid)) item={true}>
            <ForwardRef(Grid) classes={{...}} item={true}>
              <div className=\\"MuiGrid-root MuiGrid-item\\">
                <WithStyles(ForwardRef(Typography)) align=\\"center\\" color=\\"textPrimary\\" variant=\\"h3\\">
                  <ForwardRef(Typography) classes={{...}} align=\\"center\\" color=\\"textPrimary\\" variant=\\"h3\\">
                    <h3 className=\\"MuiTypography-root MuiTypography-h3 MuiTypography-colorTextPrimary MuiTypography-alignCenter\\">
                      <WithStyles(ForwardRef(Grid)) item={true}>
                        <ForwardRef(Grid) classes={{...}} item={true}>
                          <div className=\\"MuiGrid-root MuiGrid-item\\">
                            Whoops...
                          </div>
                        </ForwardRef(Grid)>
                      </WithStyles(ForwardRef(Grid))>
                      <WithStyles(ForwardRef(Grid)) item={true}>
                        <ForwardRef(Grid) classes={{...}} item={true}>
                          <div className=\\"MuiGrid-root MuiGrid-item\\">
                            Looks like you took a wrong turn
                          </div>
                        </ForwardRef(Grid)>
                      </WithStyles(ForwardRef(Grid))>
                    </h3>
                  </ForwardRef(Typography)>
                </WithStyles(ForwardRef(Typography))>
                <WithStyles(ForwardRef(Typography)) align=\\"center\\" color=\\"textPrimary\\" variant=\\"h6\\">
                  <ForwardRef(Typography) classes={{...}} align=\\"center\\" color=\\"textPrimary\\" variant=\\"h6\\">
                    <h6 className=\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorTextPrimary MuiTypography-alignCenter\\">
                      &lt; 404 Not Found &gt;
                    </h6>
                  </ForwardRef(Typography)>
                </WithStyles(ForwardRef(Typography))>
              </div>
            </ForwardRef(Grid)>
          </WithStyles(ForwardRef(Grid))>
        </div>
      </ForwardRef(Grid)>
    </WithStyles(ForwardRef(Grid))>
  </NotFound>
</ThemeProvider>"
`;
